diff --git a/configure b/configure
index 323f574..8020916 100755
--- a/configure
+++ b/configure
@@ -467,6 +467,20 @@ case $OPERATINGSYSTEM in
         fi
         echo "Configuring for Solaris (" $SYSTEM "SunOS version" $VERSION ")" 
     ;;                                                                        
+    QNX)
+        EXTRA_OBJECTS="sys_generic.o sys_qnx.o sys_posix.o"
+        EXTRA_LIBS="$LIBS -lsocket"
+        EXTRA_CLI_LIBS="$LIBS -lsocket"
+        # LIBS="$LIBS -lsocket -lnsl -lkvm -lelf -lresolv"
+        try_setsched=1
+        try_lockmem=1
+        feat_refclock=0
+        CFLAGS="-O2 -g -fPIE -fstack-protector-strong --param=ssp-buffer-size=4 -Wmissing-prototypes -Wshadow -Wall -pthread -I/usr/include/p11-kit-1"
+        add_def QNX
+        # QNX doesn't have SA_RESTART
+        add_def SA_RESTART 0
+        echo "Configuring for Blackberry (" $SYSTEM " NTO version" $VERSION ")"
+    ;;                                                                        
     * )
         echo "error: $SYSTEM is not supported (yet?)"
         exit 1
