diff --git a/sysincl.h b/sysincl.h
index 873a3bd3..f995719f 100644
--- a/sysincl.h
+++ b/sysincl.h
@@ -53,7 +53,9 @@
 #include <sys/time.h>
 #include <sys/types.h>
 #include <sys/un.h>
-#include <sys/shm.h>
+#ifndef QNX
+ #include <sys/shm.h>
+#endif 
 #include <sys/wait.h>
 #include <syslog.h>
 #include <time.h>
@@ -61,6 +63,8 @@
 
 #if defined(LINUX) || defined(FREEBSD) || defined(NETBSD) || defined(SOLARIS) || defined(HAVE_MACOS_SYS_TIMEX)
 #include <sys/timex.h>
+#elif defined(QNX)
+#include "adjtime_qnx.h"
 #endif
 
 #ifdef FEAT_IPV6

