diff --git a/configure.ac b/configure.ac
index 6a43fd5d..87f0ec78 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1950,6 +1953,10 @@ if test "$USE_NSS" = "no"; then
 			*freebsd*|*FreeBSD)
 				AC_LIBOBJ(getentropy_freebsd)
 			;;
+			*qnx*|QNX)
+                # For now base on freebsd, in the future we can have our own implementation
+				AC_LIBOBJ(getentropy_freebsd)
+			;;
 			*linux*|Linux|*)
 				AC_LIBOBJ(getentropy_linux)
 				AC_CHECK_FUNCS([SHA512_Update],,[
@@ -2235,6 +2242,16 @@ dnl includes
 
 #if STDC_HEADERS
 #include <stdlib.h>
+#if defined(__QNX__)
+/** 
+ Our stdlib.h isn't the gnu one, and it doesn't include
+ <sys/types.h which ends up importing <sys/select.h> which ends up
+ importing <sys/struct_timeval.h> (which doesn't exist on our
+ system, so we need <sys/time.h>)
+*/
+#include <sys/time.h> 
+#endif 
+
 #include <stddef.h>
 #endif
 

