diff --git a/util/config_file.c b/util/config_file.c
index 25c2b3e3..5bf3d726 100644
--- a/util/config_file.c
+++ b/util/config_file.c
@@ -210,7 +210,7 @@ config_create(void)
 	cfg->if_automatic = 0;
 	cfg->if_automatic_ports = NULL;
 	cfg->so_rcvbuf = 0;
-	cfg->so_sndbuf = 4*1024*1024;
+	cfg->so_sndbuf = 0;
 	cfg->so_reuseport = REUSEPORT_DEFAULT;
 	cfg->ip_transparent = 0;
 	cfg->ip_freebind = 0;
diff --git a/doc/example.conf.in b/doc/example.conf.in
index d151f846..aeddfb31 100644
--- a/doc/example.conf.in
+++ b/doc/example.conf.in
@@ -117,7 +117,7 @@ server:
 
 	# buffer size for UDP port 53 outgoing (SO_SNDBUF socket option).
 	# 0 is system default. Set larger to handle spikes on very busy servers.
-	# so-sndbuf: 4m
+	# so-sndbuf: 0
 
 	# use SO_REUSEPORT to distribute queries over threads.
 	# at extreme load it could be better to turn it off to distribute even.

