--- fakeroot-1.37.1.1.orig/libfakeroot.c
+++ fakeroot-1.37.1.1/libfakeroot.c
@@ -859,8 +859,19 @@
     r=next_lchown(path,owner,group);
   else
     r=0;
+#ifdef __QNX__
+  // QNX bug that chown will return EACCES instead of EPERM on permission failure
+  if(r&&(errno==EACCES)){
+    // check if its really EACCESS
+    if (faccessat(AT_FDCWD, path, F_OK, 0) == EOK)
+      r=0;
+    else
+      errno=EACCES;
+  }
+#else
   if(r&&(errno==EPERM))
     r=0;
+#endif

   return r;
 }
@@ -886,8 +897,19 @@
     r=next_lchown(path,owner,group);
   else
     r=0;
+#ifdef __QNX__
+  // QNX bug that lchown will return EACCES instead of EPERM on permission failure
+  if(r&&(errno==EACCES)) {
+    // check if its really EACCESS
+    if (faccessat(AT_FDCWD,path, F_OK, AT_SYMLINK_NOFOLLOW) == EOK)
+      r=0;
+    else
+      errno=EACCES;
+  }
+#else
   if(r&&(errno==EPERM))
     r=0;
+#endif

   return r;
 }
@@ -909,9 +931,15 @@
     r=next_fchown(fd,owner,group);
   else
     r=0;
-
+#ifdef __QNX__
+  // QNX bug that fchown will return EACCES instead of EPERM on permission failure
+  // we dont need a extra check since we already have a FD
+  if(r&&(errno==EACCES))
+      r=0;
+#else
   if(r&&(errno==EPERM))
     r=0;
+#endif

   return r;
 }
@@ -937,8 +965,19 @@
   else
     r=0;

+#ifdef __QNX__
+  // QNX bug that fchownat will return EACCES instead of EPERM on permission failure
+  if(r&&(errno==EACCES)) {
+    // check if its really EACCESS
+    if (faccessat(dir_fd, path, F_OK, flags) == EOK)
+      r=0;
+    else
+      errno=EACCES;
+  }
+#else
   if(r&&(errno==EPERM))
     r=0;
+#endif

   return r;
 }
